### PR TITLE
Fixing #14207: multiple warning about unused variable catching several cases

### LINKED ISSUE #14207: Unused variable catches more than one case is printed multiple times for unused variables that do not catch multiple cases
#### Description of the problem
```coq
Inductive type {base_type : Type} := base (t : base_type) | arrow (s d : type).
Global Arguments type : clear implicits.
Fixpoint interp {base_type} (base_interp : base_type -> Type) (t : type base_type) : Type
  := match t with
     | base t => base_interp t
     | arrow s d => @interp _ base_interp s -> @interp _ base_interp d
     end.
Axiom admit : forall {T}, T.
Section with_base.
  Context {base_type : Type}
          {base_interp : base_type -> Type}.
  Local Notation type := (@type base_type).

  Fixpoint default {t} : interp base_interp t
    := match t with
       | base x => admit
       | arrow s d => fun _ => @default d
       end.
  (* Warning: Unused variable s catches more than one case. [unused-pattern-matching-variable,pattern-matching]
Warning: Unused variable s catches more than one case. [unused-pattern-matching-variable,pattern-matching]
*)
End with_base.
```

What's up with this?  cc @herbelin 

I assume it has something to do with trying to elaborate the `match` in multiple ways, since adding `return interp _ t` suppresses the error message?

#### Coq Version

master (bba57138fcb0badf307fb00b97d73b4933e3022b)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix 

Fixes / closes #14207

The problem was caused by using side effects being iterated as many times as the number of different ways to compile the matching. The fix is by side effect too: we reset them at each new try.

- [X] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by herbelin** [comment]
Updating the PR to avoid side effects. This additionally improves situations with multiple patterns where the same variable was reported several times (see test-suite).

File: test-suite/output/Cases.out [hunk 1]
```diff
@@ -232,3 +232,12 @@ match x with
 | @D' _ _ _ _ n _ p _ => (n, p)
 end
      : J' bool (true, true) -> nat * nat
+File "./test-suite/output/Cases.v", line 313, characters 3-4:
```
Comment (ppedrot) [inline test-suite/output/Cases.out:235]: The location doesn't seem to be correct, the CI prints stdin instead.

Comment (herbelin) [reply test-suite/output/Cases.out]: Thanks, fixed. Btw, I did not find yet how to test a single file of the test-suite with dune.

**Review by ppedrot — APPROVED** [review]
Removing the mutable reference from case compilation is the the change I've always wanted to perform but never dared to do. I am surprised that it's working so easily and that there is no hidden miscompilation, but the CI seems OK so good to me.

**Comment by herbelin** [comment]
> Removing the mutable reference from case compilation is the the change I've always wanted

Note that this pattern-matching compilation algorithm is in the process of being replaced by a new one implementing small inversion at all depths (and thus basically implementing "dependent pattern matching")

**Comment by JasonGross** [comment]
> Note that this pattern-matching compilation algorithm is in the process of being replaced by a new one implementing small inversion at all depths (and thus basically implementing "dependent pattern matching")

That's exciting!  Is this work public / is there a place I can peek in on the progress?

**Comment by herbelin** [comment]
This is [here](https://github.com/thierry-martinez/coq/tree/small_inversion_thierry) with a couple of examples now working [here](https://github.com/thierry-martinez/small_inversion/blob/master/test.v), starting with `Definition tail2 A n (v : vector A (S (S n))) := match v with Cons _ _ _hd tl => tl end.`, etc.

Incidentally, the algorithm exploits OCaml's GADT to ensure that the terms are manipulated in their right typing context (remembering the number of declarations in the context for static correctness of de Bruijn indices).

The stdlib compatibilty is almost obtained (three issues remaining), but knowing how the current algorithm is already sensitive to small changes, it is not sure that we will be able to have a 100% compatibility. I believe the price is worth though.

**Pull request merged**